File: daily_problems/2024/10/1005/personal_submission/cf558d_yawn_sean.cpp
#include <bits/stdc++.h>
// #pragma GCC optimize("O3,Ofast,unroll-loops")
// #include <bits/extc++.h>
// using namespace __gnu_cxx;
// using namespace __gnu_pbds;
using namespace std;

signed main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    int h, q;
    cin >> h >> q;

    map<long long, int> mp;
    mp[1ll << h - 1] = 0;
    mp[1ll << h] = 0;

    for (int i = 0; i < q; i ++) {
        int idx, ans;
        long long l, r;
        cin >> idx >> l >> r >> ans;
        r ++;
        l <<= h - idx;
        r <<= h - idx;

        if (ans) mp[l] ++, mp[r] --;
        else mp[1ll << h - 1] ++, mp[l] --, mp[r] ++, mp[1ll << h] --;
    }

    int cur = 0;
    long long start = 0, cnt = 0, last = 1ll << h  - 1;

    for (auto &[k, v]: mp) {
        if (cur == q) {
            cnt += k - last;
            start = last;
        }
        cur += v;
        last = k;
    }

    if (cnt == 0) cout << "Game cheated!";
    else if (cnt == 1) cout << start;
    else cout << "Data not sufficient!";

    return 0;
}

File: daily_problems/2024/10/1005/personal_submission/cf558d_yawn_sean.py
# Submission link: https://codeforces.com/contest/558/submission/284379794
def main():
    h, q = MII()
    ls = [1 << h - 1]
    rs = [1 << h]

    for _ in range(q):
        i, l, r, ans = LII()
        r += 1
        l <<= h - i
        r <<= h - i
        
        if ans:
            ls.append(l)
            rs.append(r)
        else:
            ls.append(1 << h - 1)
            rs.append(l)
            ls.append(r)
            rs.append(1 << h)

    pts = sorted(set(sorted(ls + rs)))
    cnt = Counter()

    for pt in pts:
        cnt[pt] = 0

    for l in ls:
        cnt[l] += 1

    for r in rs:
        cnt[r] -= 1

    cur = 0
    freq = 0
    start = 0
    for i in range(len(pts) - 1):
        cur += cnt[pts[i]]
        if cur == q + 1:
            freq += pts[i + 1] - pts[i]
            start = pts[i]

    if freq == 0: print('Game cheated!')
    elif freq == 1: print(start)
    else: print('Data not sufficient!')